File: src/main/java/me/testaccount666/serversystem/userdata/money/AbstractBankAccount.java
package me.testaccount666.serversystem.userdata.money;

import lombok.Getter;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Map;
import java.util.UUID;

@Getter
public abstract class AbstractBankAccount {
    protected final UUID owner;
    protected final BigInteger accountId;

    public AbstractBankAccount(UUID owner, BigInteger accountId) {
        this.owner = owner;
        this.accountId = accountId;
    }

    public abstract BigDecimal getBalance();

    public abstract void setBalance(BigDecimal balance);

    public abstract void delete();

    public abstract void save();

    public void deposit(BigDecimal amount) {
        setBalance(getBalance().add(amount));
    }

    public void withdraw(BigDecimal amount) {
        setBalance(getBalance().subtract(amount));
    }

    public void transfer(BigDecimal amount, AbstractBankAccount targetAccount) {
        withdraw(amount);
        targetAccount.deposit(amount);
    }

    public boolean hasEnoughMoney(BigDecimal amount) {
        return getBalance().compareTo(amount) >= 0;
    }

    public abstract Map<UUID, BigDecimal> getTopTen();
}

File: src/main/kotlin/me/testaccount666/serversystem/userdata/money/AbstractBankAccount.kt
package me.testaccount666.serversystem.userdata.money

import java.math.BigDecimal
import java.math.BigInteger
import java.util.*

abstract class AbstractBankAccount(val owner: UUID, val accountId: BigInteger) {
    abstract var balance: BigDecimal

    abstract fun delete()

    abstract fun save()

    fun deposit(amount: BigDecimal) {
        balance += amount
    }

    fun withdraw(amount: BigDecimal) {
        balance -= amount
    }

    fun transfer(amount: BigDecimal, targetAccount: AbstractBankAccount) {
        withdraw(amount)
        targetAccount.deposit(amount)
    }

    open fun hasEnoughMoney(amount: BigDecimal): Boolean = balance >= amount

    abstract val topTen: MutableMap<UUID, BigDecimal>
}